File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.cpp
#include <string>
#include <vector>
// 그냥 bfs나 dfs로 하면 막혔던 길을 거슬러 올라가 다시 시도하지 않음. 
// 중복해서 방문하면 양과 늑대가 추가되지 않음. 
// root노드에서 전체적으로 depth가 높아지면서 최대한 많은 양들을 모으면 됨!

using namespace std;
int answer;
vector<vector<int>> adj;
vector<int> candidate; 

void dfs(vector<int>& info, vector<int> candidate, int sheep, int wolf){
    for(int i = 0; i < candidate.size(); i++){    
        int node = candidate[i]; 
        int sheepNum = info[node] == 0 ? sheep+1 : sheep; 
        int wolfNum = info[node] == 1 ? wolf +1 : wolf;
        answer = max(answer, sheepNum);
        
        // 이동이 불가능(늑대 >= 양)이 되면 다른 후보지 선택 
        // -> 전부 불가하면 다음 턴으로 넘어가지만, 후보지군은 줄어들음. 
        if(wolfNum >= sheepNum) continue;
        
        // 이동 가능하면 후보지군 늘리기, 기존에서 자신 뺀 리스트도 더하기 
        vector<int> next_candidate = candidate; 
        next_candidate.erase(next_candidate.begin()+i);
        for(auto& child: adj[node]){
            next_candidate.push_back(child);
        }
            
        // 다음 턴으로 이동
        dfs(info, next_candidate, sheepNum, wolfNum);
        
    }
}
int solution(vector<int> info, vector<vector<int>> edges) {
    // 전역변수 초기화
    answer = 0; 
    adj.assign(info.size(), vector<int>());
    
    // 인접리스트 만들기 
    for(auto& e: edges){
        int a = e[0]; int b = e[1];
        adj[a].push_back(b); 
    }
    
    // 후보군 중 하나 골라 가기
    candidate.push_back(0);
    dfs(info, {0}, 0, 0);
    
    // 모을 수 있는 양 최대 수
    return answer;
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*; 


class Solution {
    int maxSheep;
    int[] info;
    List<Integer>[] graph; // 연결 그래프(부모-자식)
    
        
    public int solution(int[] info, int[][] edges) { // edges: [부모 노드 번호, 자식 노드 번호]
        this.info = info;
        this.maxSheep = 0;
        
        // 연결 그래프 만들기
        graph = new ArrayList[info.length];
        for(int i = 0; i < info.length; i++) graph[i] = new ArrayList<>(); 
        for(int[] e : edges){
            int parent = e[0]; int child = e[1]; 
            graph[parent].add(child); 
        }
        
        // 각 상황에 존재하는 양과 늑대 수
        List<Integer>[] animals = new ArrayList[2]; // 양, 늑대 수
        for(int i = 0; i < 2; i++) animals[i] = new ArrayList<>(); 
        animals[0].add(0); // 첫 양의 위치 삽입
        
        // 백트래킹으로 검색 시작
        gatherAnimals(0, 0, animals);

        // 모을 수 있는 양은 최대 몇 마리인지
        return maxSheep;
    }
    
    public void gatherAnimals(int sheepNum, int wolfNum, List<Integer>[] animals){ // 양, 늑대 누적 수 
        
        // 양이 있다면 다 데려오기(답 갱신) -> 그의 자식들 전부 배열에 넣어주기
        while(!animals[0].isEmpty()){
            int sheepNode = animals[0].remove(animals[0].size() - 1); // 제거하며 반환
            sheepNum++;
            addChildren(sheepNode, animals); // 자식 추가

            maxSheep = Math.max(maxSheep, sheepNum);
        }

        // 늑대 데려갈 수 없다면 return 
        if(sheepNum <= wolfNum + 1) return;

        // 양이 없다면 늑대 한 마리씩 차례대로 돌아가면서 데려가기
        for (int i = 0; i < animals[1].size(); i++) {
            int wolfNode = animals[1].get(i); // 늑대 한 마리 선택

            List<Integer>[] nextAnimals = copyAnimals(animals); // 얕은 복사!!
            nextAnimals[1].remove(i); // 이번 늑대는 방문했으니 제거
            addChildren(wolfNode, nextAnimals); // 자식 추가
            gatherAnimals(sheepNum, wolfNum+1, nextAnimals); // 재귀 (새로 만든 배열 전달해서 이전의 데이터는 유지되도록)
        }
            
    }

    // 자식을 배열에 추가해주기 (양/늑대 구분)
    private void addChildren(int node,List<Integer>[] animals) {
        for (int child : graph[node]) {
            if (info[child] == 0) {
                animals[0].add(child);
            } else {
                animals[1].add(child);
            }
        }
    }
    
    // 안의 내용까지 얕은 복사
    private List<Integer>[] copyAnimals(List<Integer>[] animals) {
        List<Integer>[] copied = new ArrayList[2];

        copied[0] = new ArrayList<>(animals[0]);
        copied[1] = new ArrayList<>(animals[1]);

        return copied;
    }
}